### PR TITLE
kvclient(ticdc): close the grpc client after all goroutine exited to prevent data race and panic

### LINKED ISSUE #10799: cdc panic: runtime error: invalid memory address or nil pointer dereference
### What did you do?

This panic is seen after https://github.com/pingcap/tiflow/issues/10798

### What did you expect to see?

cdc no panic

### What did you see instead?

```
{"level":"warn","ts":"2024-03-17T03:44:29.64617Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0038b4000/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
{"level":"warn","ts":"2024-03-17T03:44:35.647313Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0038b4000/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
{"level":"warn","ts":"2024-03-17T03:44:41.648666Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0038b4000/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
{"level":"warn","ts":"2024-03-17T03:46:41.939417Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc003b921c0/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x3878de9]

goroutine 9470229 [running]:
github.com/pingcap/tiflow/cdc/kv.(*requestedStream).receive(0xc0706b43c0, {0x5c10998, 0xc0706b52c0}, 0xc0747ea580, 0xc06e0f2980, 0x1959945?, 0xc0706b48c0?)
        github.com/pingcap/tiflow/cdc/kv/shared_stream.go:225 +0x69
github.com/pingcap/tiflow/cdc/kv.(*requestedStream).run.func3()
        github.com/pingcap/tiflow/cdc/kv/shared_stream.go:187 +0x2e
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync@v0.6.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 9470087
        golang.org/x/sync@v0.6.0/errgroup/errgroup.go:75 +0x96
```

### Versions of the cluster

cdc version: 
cdc version:
Release Version: v8.0.0                                  
Git Commit Hash: 130403f3b9b8ad8a28ceada642277986e317ebc2
Git Branch: heads/refs/tags/v8.0.0                       
UTC Build Time: 2024-03-15 13:58:58                      
Go Version: go version go1.21.6 linux/amd64              
Failpoint Build: false

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #10799 

### What is changed and how it works?

* release grpc client after all goroutines existed.

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
`None`
```

### CODE REVIEW

**Comment by flowbehappy** [comment]
Required UT

File: cdc/kv/shared_stream.go [hunk 1]
```diff
@@ -196,13 +202,23 @@ func (s *requestedStream) run(ctx context.Context, c *SharedClient, rs *requeste
 
 		s.tableExclusives = make(chan tableExclusive, 8)
 		g.Go(func() error {
+			connAndClientsCache := make(map[SubscriptionID]*sharedconn.ConnAndClient)
+			defer func() {
+				for _, cc := range connAndClientsCache {
```
Comment (CharlesCheung96) [inline cdc/kv/shared_stream.go:207]: Maybe the connections held by `s.tableExclusives` should also be released?

File: cdc/kv/shared_stream.go [hunk 2]
```diff
@@ -196,13 +197,32 @@ func (s *requestedStream) run(ctx context.Context, c *SharedClient, rs *requeste
 
 		s.tableExclusives = make(chan tableExclusive, 8)
 		g.Go(func() error {
+			connAndClientsCache := make(map[SubscriptionID]*sharedconn.ConnAndClient)
```
Comment (asddongmen) [inline cdc/kv/shared_stream.go:200]: Please add a comment to indicate this cached is only use to release conn when this goroutine exits.

**Review by asddongmen — APPROVED** [review]

**Pull request closed without merging**